Propose new channel communication protocol spec (#3530)
project-oak/oak: oak_channel/SPEC.md

# Oak Restricted Kernel Communication Protocol Specification

This document serves as a specification of the protocol used for communication
between the Untrusted Launcher (running on the host) and Enclave Binry built
with the Restricted Kernel (running inside a guest VM).

In the specification below the Enclave Binary will be referred to the as the
"service". The Untrusted Launcher will be referred to the as the "client".

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119).

## Layer Defintions

Conceptually, the Oak Restricted Kernel Communication Protocol is made up of

> three layers. The [Invocation Layer](#invocation-layer) uses protocol buffers

Reviewer: It is not clear to me what is the lowest layer of the stack here. From the description I expect the message layer to be at the highest level of the stack but the fact that is stated in the middle makes me uncertain about that.

Reviewer 2: I found this a bit hard to read as well; consider making it a bulleted list.

(protobuf) to define the methods exposed by the service. It is responsible for
encoding requests and responses that wrap input parameters and return values.
The [Message Layer](#message-layer) provides a logical wrapper for the encoded
requests and responses. The [Frame Layer](#frame-layer) converts messages into
frames which are then sent over the transport channel.

> ### Invocation Layer

Reviewer 3: How does this relate to `oak_idl`?

> Methods define the functionality exposed by service. Each method defines input

Reviewer: ```suggestion
Methods define the functionality exposed by the service. Each method defines input
```

parameters and return types. Protocol buffers (protobuf) is used to define which

> methods are available, the method ids, input parameters and return types. It

Reviewer 3: It uses ... -> The service converts the requests and responses into bytes using standard protobuf encoding.

uses standard protobuf encoding to convert the requests and responses into

> bytes.

Reviewer 3: The "bytes" are the request messages/response message---true?

#### Invocation

> Each method invocation consists of two parts: a request message sent by the

Reviewer 4: ```suggestion
Each method invocation consists of two messages: a request message sent by the
```

messages follows an identical format.

#### Message

The invocation ID and message length is already encoded into the frame header,
so an encoded message does not have a header of its own and MUST only consist of
a body containing the encoded bytes of the request or response:

```text
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                         body bytes...                         +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
```

<!-- Diagram generated with https://www.luismg.com/protocol/, using the spec
"body bytes...:64"  -->

### Frame Layer

Messages are sent and received as frames with a maximum total length of 4,096
bytes per frame. Each message is fragmented into a set of one or more frames.

#### Frame

MUST consist of a header and a body. The encoded header is 16 bytes long and
MUST consist of the following fields:

> - `protocol_version`, unsigned 16-bit little-endian integer

Reviewer 5: Great idea to start thinking about versioning here.

  The version of the communication protocol in use. This value MUST be 1 for
  this version of the specification.

> - `body_length`, unsigned 16-bit little-endian integer

Reviewer 5: Wouldn't it be better to have the total frame length here instead of just the body? That way, if we evolve the protocol to a future version, as long as we keep these two fields the same, a parser would be able to at least delimit the entire frame, even if it cannot parse it. At the moment, the parser would have to know the exact length of all the fields for the specific version. Also, even if we have only one version, it seems nice to logically separate the delimitation of frames via their length from their actual parsing. WDYT?

+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                          frame_number                         |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                         invocation_id                         |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                         body bytes...                         +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
```

<!-- Diagram generated with https://www.luismg.com/protocol/, using the spec
"protocol_version:16,body_length:16,message_length:32,frame_number:32,invocation_id:32,body bytes...:64" -->

The maximum total length of a single frame is 4,096 bytes.

## Interaction between the Layers

### Sending Messages

To send a message, the sender MUST first encode the message into frames:

The sender MUST chunk byte representation of the message into a set of frame
bodies. The last frame body of the resulting set MAY be shorter than 4,080 bytes
(the maximum length of a frame body). Any preceding frame bodies MUST be exactly

> 4,080 bytes in length. The sender MUST then create a frame from each frame body.

Reviewer 5: I would like to leave it to the client to decide what framing size to use, and just set the upper bound. And we should handle frames of different sizes within the same message too IMO.

Author: If full frames don't have a fixed size the sender will need to put the number of frames and the message size in the header if we want to be able to reserve the appropriate amount or memory and know when we have received all the frames. I will add an additional field to the header to support that.

Reviewer 5: Isn't the message size (which is already there) enough?

Author: It is enough to reserve the space, but since we are not guaranteed that each full frame will have the same size we cannot calculate the expected number of frames from the message size.

Reviewer 5: But I thought in practice the logic on the receiving side is to allocate the space, then append each frame in order until it's full. We don't need to know the number of frames ahead of time, just check that the total size adds up correctly.

Reviewer 5: Let's step back for a moment. Is the issue that we risk missing a frame within a message and not notice? Or that individual bits may be corrupted? If the former, then a flag indicating the last frame of a message may be enough. For the latter, I agree some sort of checksum may be useful, though that may be also done at the application layer in principle.

Author: The important part is to have very high confidence that we will detect frame misalignment or corruption of the header, as this will indicate an unrecoverable situation for the channel and we should terminate the workload. Knowing an exact expected value (or small valid range) for each header field for a specific frame helps a lot with this.

If any bits in the body are corrupted we will detect it (at least for encrypted user requests) when we try to decrypt it, but that should only affect the single message that is corruped, not the stream as a whole, so we won't have to terminate the workload.

In my previous message I mentioned a checksum for the message, but actually meant a checksum for the frame.

Reviewer 5: Makes sense. So perhaps have a small checksum and also flags for each header? similar to TCP.

Reviewer 2: +1, I'd like to see some kind of checksum. Memory corruption, even inside one machine, is all too common.

Author: I don't think out-of-order frames or entire frames going missing is likely to happen due to data corruption, so we probably don't need frame numbers if we have a checksum to validate the header of the frame and rely on message size to know when we have received everything.

If we leave message integrity up to the application (e.g. since we are using authenticated encryption for user requests) we only really need the checksum to cover the header:

- Version
- Frame size
- Message Size
- Invocation ID
- Header Checksum

The header checksum could be something simple like the result or xoring previous batches of 4 bytes from the header, or it could be a truncated hash.

WDYT?

##### Handling an invocation

> The service MUST respond to each request message with a single response message.

Reviewer 5: ```suggestion
The service MUST respond to each request message with exactly one response message.
```

The service MAY send a response message before the request message has been

> received in its entirety, but MUST NOT send start sending a response before any

Reviewer 3: IIUC "it its entirety" implies that messages be only partially sent/received, i.e., not atomic.

If so, what does: "Sending a response message" mean excatly below? The point where the service starts sending a message or the entire message is sent?

Author: The point when the complete message has been transmitted.

> received in its entirety, but MUST NOT send start sending a response before any

Reviewer: ```suggestion
received in its entirety, but MUST NOT start sending a response before any
```

Reviewer: How could a response message be sent before any part of the request has been received? How would the server know where to send the response to? Or is a connection already made at this stage?

> part of the request message has been received. Sending a response message

Reviewer 4: I'm not sure if the response should be called a message, since this is the invocation layer and not the message layer.
I think it's better to call it "Invocation request" and "Invocation response". And keep the "message" only for describing the "Message layer" entities.

concludes the invocation.

The client MUST NOT send response messages.

### Message Layer

Messages are logical units used to transport requests and responses.

> There are exactly two messages types: request messages and response messages.

Reviewer: ```suggestion
There are exactly two message types: request messages and response messages.
```

Since the invocation layer is responsible for encoding requests and reponses, at

> the communication protocol level the encoding for request and response messages

Reviewer 3: nit: encoding for request messages.

> follow an identical format.

Reviewer 5: Worth mentioning that requests will always "flow" in one direction, and responses in the opposite direction? I.e. that each logical channel has fixed roles for client and server, and invocations are always initiated by the client.

> follow an identical format.

Reviewer: ```suggestion
follows an identical format.
```

#### Message

> A byte encoded message MUST consist only of the encoded bytes:

Reviewer 5: +1 this is a bit surprising (though I understand the point is to specify that there is no length prefix here)

> A byte encoded message MUST consist only of the encoded bytes:

Reviewer 2: As proposed, the message layer indeed sounds completely unnecessary. It just states that "requests and responses are sequences of bytes", which is pretty much always the case.

Why do we need this layer at all?

 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                         invocation_id                         |

> +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

Reviewer 5: Maybe a nit, but I think it would be nice to keep length as the first field for consistency and to simplify the low level parsing logic, and also potentially allow stacking combinators to express this transformation (e.g. receive byte stream, cut it into frames by just looking at the length field and send the content to the next combinator, etc.). We don't have to actually implement it that way, but I think it's nice if it could be done from the outside in if desired.

Author: I think we will always parse the entire header for a layer before parsing the bytes, so the exact order doesn't matter. But I have swapped the body length and the invocation ID so the two length fields are adjacent.

> |                         message_length                        |

Reviewer 5: I'm confused -- is this really the higher-level message length, or just this frame length?

If the high-level message length, I'm not sure why it needs to be here.
If the frame length, it seems redundant with the body length; in fact I think it would be better to have the overall frame length instead of just the body length.

Reviewer 4: I think the reason for putting the message length here was to preallocate a correct amount of bytes for the whole message once the first frame is received.

Author: This is really the message length. There are two reasons for storing the message length in the frame header. The first is as Ivan pointed out is to be able to reserve memory for the message so that we don't have to keep reallocating when we receive a new frame. The second is for frame validation.

Regarding the overall frame length vs the body length: we need both so we have to do a calculation anyway to determine the one that is not specified. The reason for choosing to specify the body length rather than the frame length is that it matches the approach in the virtio spec.

+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                          frame_number                         |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                          body_length                          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                         body bytes...                         +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
```

<!-- Diagram generated with https://www.luismg.com/protocol/, using the spec
"invocation_id:32,message_length:32,frame_number:32,body_length:32,body bytes...:64" -->

> - `invocation_id`, u32, little endian

Reviewer 2: Instead of specifying that all these fields are "little endian", I'd propose "machine native endianness" -- these messages will be exchanged between processes on the same machine, so if we ever move to a BE machine having to do two swaps seems a bit silly.

Reviewer 4: Does the restricted kernel's endiannes also depend on the architecture?

Reviewer 2: Yes, on a big-endian CPU we'd run the kernel in big-endian mode as well.

Author: I disagree. Communication protocol specifications are usually explicit about endianness to avoid doubt. Also, at the moment we are running the untrusted launcher and the host on the same machine, but that might not always be the case so we cannot assume that we will always have the same endianness on both sides of the channel (even though it seems extremely unlikely that we will ever run either part on a BE machine).

>   The unique ID of the invocation that the frame's forms a part of. All the

Reviewer: I don't understand this sentence.

Author: A message is split into frames. All these frames that make up a message must have the same invocation ID for that message. The invocation ID is used to know which message a frame belongs to.

  frames that make up a message MUST have the same invocation ID. The invocation
  ID of a response message MUST match the value of its related request message.
  The invocation ID MUST be incremented for each method invocation, naturally
  wrapping when the maximum value for the data type is reached.

- `message_length`, unsigned 32-bit little-endian integer

>   The length of the overall message. This value MUST be larger or equal to

Reviewer 4: Is it about the upper level message length or the frame length?

Author: The "message length" is the length of the message, not the frame.

  `body_length`. The message length for all frames that make up the message MUST
  be the same.

  The size of a single message MUST NOT exceed 2GiB.

> - `frame_number`, unsigned 32-bit little-endian integer

Author: This can be calculated from the message size.

>   The position of this frame within the overall message. The first frame MUST

Reviewer 4: Does the channel implementation need to assume that all of the frames in the "stream" will be maximum 4K length (in order to calculate the abount of expected frames) ?

Author: Yes, this is specified lower down under "Sending messages"

  start at number 0. Each frame's number MUST be 1 higher than the previous
  frame. Frames for a single message MUST be sent in ascending order of the
  frame number.

- `body_length`, unsigned 32-bit little-endian integer

>   The length of the body byte array of the rfram. This value must not exceed

Author: Done.

>   4,080. The sum of the body lengths of all the frames in a message must equal

Author: This is a number, so in my opinion the maximum value should be specified as a number, not a number of bytes.

We have a maximum frame size of 4096 bytes and the header is 16 bytes, so 4080 is what is left over.

  the message length.

- `body`, variable length byte array

  Byte fragment of an encoded message. This size of the byte array MUST NOT
  exceed 4,080 bytes.

> The maximum total length of a single frame is 4,096 bytes.

Author: This is a choice. It happens to match the buffer size we use for some of our comms channels over shared memory and is the default size for memory pages, but is still an arbitrary choice.

## Interaction between the Layers

### Sending Messages

> To send a message, it MUST first be encoded into frames.

Reviewer 3: nit: Can we formulate to active? Is it the "Sender" who has to do this?

To send a message, the sender MUST encode a message into frames. The sender MUST chunk the byte representation of a message into a set of frame bodies. [..]

The sender must form a frame from each frame body and MUST send the resulting frame set in sequenctial order of the frame number over the channel.

Reviewer 5: +1

Author: Done.

The byte representation of a message MUST be chunked into a set of frame bodies.
The last frame body of the resulting set MAY be shorter than 4,080 bytes (the
maximum length of a frame body). Any preceding frame bodies MUST be exactly

> 4,080 bytes in length.

Reviewer 5: Is this precluding the option of having smaller frames then? That may be problematic if in the future we allow larger frame sizes, since there would be no backwards compatibility. But perhaps we don't care about it to keep things simple.

Author: Yes. The alternative is to also specify the expected number of frames. I would rather add a protocol version and fix the frame size per protocol version rather than allowing different frame sizes withing the same protocol.

A frame MUST then be formed from each frame body. The resulting frame set MUST
be sent in sequential order of the frame number over the channel.

### Receiving Messages

To receive a message the recipent incrementally reads frames from the
communication channel. Messages are parsed from frames by appending the frame

> bodies in the order they are received. The recipient SHOULD validate the

Reviewer 3: Why SHOULD the recipient validate the sequence?

Reviewer 5: +1 I would expect this to be MUST, and in fact the client should abort if an out-of-order frame is received.

Author: I disagree. I interpret MUST as something that will always break the protocol if it is ignored. Not validating this will not detect corruption if it happened, but it will not actively break correctly formatted things.

Reviewer 5: `MUST   This word, or the terms "REQUIRED" or "SHALL", mean that the
   definition is an absolute requirement of the specification.`

`SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.`

(From https://www.rfc-editor.org/rfc/rfc2119)

I think this should be an absolute requirement of the specification, so MUST is appropriate.

Author: I will change it to MUST, but I think SHOULD could also be appropriate. In the case where the communcations channel is infallible and the sender is trusted to accurately follow the protocol it could be a valid decision to skip this validation.
